File: frontend/src/api/color.ts
import type {
  ColorClassifierConfig,
  ColorClassifySampleRequest,
  ColorDecisionResult,
  ColorCalibrationStatus,
  MaskPreviewResult,
} from '../types/color'

function apiBaseUrl(): string {
  const configured = import.meta.env.VITE_BACKEND_API_URL as string | undefined
  if (configured) return configured.replace(/\/$/, '')
  if (window.location.port && window.location.port !== '5173') return window.location.origin
  return `${window.location.protocol}//${window.location.hostname}:8000`
}

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await fetch(`${apiBaseUrl()}${path}`, {
    headers: { 'Content-Type': 'application/json', ...init?.headers },
    ...init,
  })
  const body = await response.json()
  if (!response.ok) throw new Error(JSON.stringify(body))
  return body as T
}

export function fetchColorConfig(): Promise<ColorClassifierConfig> {
  return request<ColorClassifierConfig>('/api/color/config')
}

export function updateColorConfig(config: ColorClassifierConfig): Promise<ColorClassifierConfig> {
  return request<ColorClassifierConfig>('/api/color/config', {
    method: 'PUT',
    body: JSON.stringify(config),
  })
}

export function classifyColorSample(requestBody: ColorClassifySampleRequest): Promise<ColorDecisionResult> {
  return request<ColorDecisionResult>('/api/color/classify-sample', {
    method: 'POST',
    body: JSON.stringify(requestBody),
  })
}

export function fetchLatestColorDecision(): Promise<ColorDecisionResult | null> {
  return request<ColorDecisionResult | null>('/api/color/latest')
}

export function resetColor(): Promise<{ reset: boolean }> {
  return request<{ reset: boolean }>('/api/color/reset', { method: 'POST' })
}

export function fetchColorCalibration(): Promise<ColorCalibrationStatus> {
  return request('/api/color/calibration')
}

export function addColorCalibrationReference(payload: { expected_team: 'enemy' | 'friend'; capture_id: string }): Promise<ColorCalibrationStatus> {
  return request('/api/color/calibration/references', { method: 'POST', body: JSON.stringify(payload) })
}

export function resetColorCalibration(): Promise<ColorCalibrationStatus> {
  return request('/api/color/calibration/reset', { method: 'POST' })
}

export function previewColorMask(requestBody: ColorClassifySampleRequest): Promise<MaskPreviewResult> {
  return request<MaskPreviewResult>('/api/color/preview-mask', {
    method: 'POST',
    body: JSON.stringify(requestBody),
  })
}

File: backend/app/api/color.py
from fastapi import APIRouter, Depends, HTTPException

from app.api.deps import get_runtime
from app.schemas.color import (
    ColorClassifierConfig,
    ColorCalibrationReferenceRequest,
    ColorCalibrationStatus,
    ColorClassifySampleRequest,
    ColorDecisionResult,
    MaskPreviewResult,
)
from app.services.runtime_state import RuntimeState

router = APIRouter(prefix="/api/color", tags=["color"])


@router.get("/config", response_model=ColorClassifierConfig)
def get_color_config(runtime: RuntimeState = Depends(get_runtime)) -> ColorClassifierConfig:
    return runtime.color_classifier.get_config()


@router.put("/config", response_model=ColorClassifierConfig)
def update_color_config(
    update: ColorClassifierConfig,
    runtime: RuntimeState = Depends(get_runtime),
) -> ColorClassifierConfig:
    if runtime.stage3_competition_profile_locked():
        raise HTTPException(status_code=409, detail="A3_PROFILE_LOCKED")
    return runtime.color_classifier.update_config(update)


@router.get("/calibration", response_model=ColorCalibrationStatus)
def color_calibration_status(runtime: RuntimeState = Depends(get_runtime)) -> ColorCalibrationStatus:
    return runtime.color_classifier.calibration_status()


@router.post("/calibration/references", response_model=ColorCalibrationStatus)
def add_color_calibration_reference(
    request: ColorCalibrationReferenceRequest,
    runtime: RuntimeState = Depends(get_runtime),
) -> ColorCalibrationStatus:
    if runtime.stage3_competition_profile_locked():
        raise HTTPException(status_code=409, detail="A3_PROFILE_LOCKED")
    try:
        return runtime.color_classifier.record_calibration_reference(request)
    except ValueError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc


@router.post("/calibration/reset", response_model=ColorCalibrationStatus)
def reset_color_calibration(runtime: RuntimeState = Depends(get_runtime)) -> ColorCalibrationStatus:
    if runtime.stage3_competition_profile_locked():
        raise HTTPException(status_code=409, detail="A3_PROFILE_LOCKED")
    return runtime.color_classifier.reset_calibration()


@router.post("/classify-sample", response_model=ColorDecisionResult)
def classify_sample(
    request: ColorClassifySampleRequest,
    runtime: RuntimeState = Depends(get_runtime),
) -> ColorDecisionResult:
    return runtime.color_classifier.classify_sample(request)


@router.get("/latest", response_model=ColorDecisionResult | None)
def get_latest_color_decision(runtime: RuntimeState = Depends(get_runtime)) -> ColorDecisionResult | None:
    return runtime.color_classifier.latest()


@router.post("/reset")
def reset_color(runtime: RuntimeState = Depends(get_runtime)) -> dict[str, bool]:
    return runtime.color_classifier.reset()


@router.post("/preview-mask", response_model=MaskPreviewResult)
def preview_mask(
    request: ColorClassifySampleRequest,
    runtime: RuntimeState = Depends(get_runtime),
) -> MaskPreviewResult:
    return runtime.color_classifier.preview_mask(request)
